include/circt/Dialect/FIRRTL/FIRRTLIntrinsics.h

```diff
@@ -13,105 +13,228 @@
+  ParseResult typedOutputElement(unsigned n, StringRef name) {
+    return checkOutputElement(n, name, [](auto ty) { return isa<T>(ty); });
+  }
+
+  template <typename T>
+  ParseResult sizedOutputElement(unsigned n, StringRef name, int32_t size) {
+    return checkOutputElement(n, name, "not size " + Twine(size),
+                              [size](auto ty) {
+                                auto t = dyn_cast<T>(ty);
+                                return t && t.getWidth() == size;
+                              });
+  }
+};
+
+/// Base class for Intrinsic Converters.
+///
+/// Intrinsic converters contain validation logic, along with a converter
+/// method to transform generic intrinsic ops to their implementation.
+class IntrinsicConverter {
+public:
+  virtual ~IntrinsicConverter() = default;
+
+  /// Checks whether the intrinsic is well-formed.
+  ///
+  /// This or's multiple ParseResults together, returning true on failure.
+  virtual bool check(GenericIntrinsic gi) = 0;
+
+  /// Transform the intrinsic to its implementation.
+  virtual void convert(GenericIntrinsic gi, GenericIntrinsicOpAdaptor adaptor,
+                       PatternRewriter &rewriter) = 0;
+};
+
+template <typename OpTy>
+class IntrinsicOpConverter : public IntrinsicConverter {
+public:
+  /// Transform the intrinsic to its implementation.
+  /// Handles the simple case of just forwarding to new op kind.
+  void convert(GenericIntrinsic gi, GenericIntrinsicOpAdaptor adaptor,
+               PatternRewriter &rewriter) final {
+    // Pass along result type and operands.  No attributes.
+    rewriter.replaceOpWithNewOp<OpTy>(gi.op, gi.op.getResultTypes(),
+                                      adaptor.getOperands());
+  }
 };
 
 /// Lowering helper which collects all intrinsic converters.
 class IntrinsicLowerings {
+public:
+  using ConversionMapTy =
+      llvm::DenseMap<StringAttr, std::unique_ptr<IntrinsicConverter>>;
+
 private:
-  using ConverterFn = std::function<LogicalResult(FModuleLike)>;
+  using ConverterFn = std::function<LogicalResult(GenericIntrinsicOp)>;
 
   /// Reference to the MLIR context.
   MLIRContext *context;
 
-  /// Reference to the instance graph to find module instances.
-  InstanceGraph &graph;
-
   /// Mapping from intrinsic names to converters.
-  DenseMap<StringAttr, ConverterFn> intmods;
-  /// Mapping from extmodule names to converters.
-  DenseMap<StringAttr, ConverterFn> extmods;
+  ConversionMapTy conversions;
 
 public:
-  IntrinsicLowerings(MLIRContext *context, InstanceGraph &graph)
-      : context(context), graph(graph) {}
+  IntrinsicLowerings(MLIRContext *context) : context(context) {}
 
   /// Registers a converter to an intrinsic name.
   template <typename T>
   void add(StringRef name) {
-    addConverter<T>(intmods, name);
-  }
-
-  /// Registers a converter to an extmodule name.
-  template <typename T>
-  void addExtmod(StringRef name) {
-    addConverter<T>(extmods, name);
+    addConverter<T>(name);
   }
```

Comment:
Could we make `add` do the replacement of `.` to `_` in the intrinsic name and call `addConverter` again with it? That way you'd only have to register an intrinsic lowering once with its proper `circt.foo.bar_baz` name, and you'd get the `circt_foo_bar_baz` by convention.

Reply:
Good point, thanks!

I wonder if we want to support both anymore anyway?  Being slightly unsure, left this as-is for now.

The new syntax could support periods in the intrinsic name but it's simpler and maybe better to have that just be a single (identifier) token -- in practice we only use the non-underscore version in documentation (see below PR) and in some internal tests.  This isn't currently reachable, all paths must use the underscores presently.

But yeah this does seem like a silly thing to push onto users of this class, as well as causes typo-prone name duplications.

cc https://github.com/llvm/circt/pull/6860 .

Reply:
Yeah good point. I guess this is more or less inspired by LLVM's use of `.` as grouping/namespacing character in its intrinsics, but then needing a workaround since Chisel/FIRRTL can't generate that 😁